Serve /crate/latest/crate rather than redirect

jsha opened the pull request:
Part of #1438

jsha commented on src/web/mod.rs:
```diff
@@ -345,14 +345,24 @@ fn match_version(
+        Some((version, id)) if req_version == "*" => {
+            return Ok(MatchVersion {
+                corrected_name,
+                version: MatchSemver::Exact((version.to_string(), *id)),
```
Changing `version` here to "latest" results in a 500 Internal Server Error for http://localhost:3000/regex/latest/regex/, with this stack trace:

```
2021/10/29 21:43:46 [ERROR] docs_rs::utils: called `cexpect!()` on a `None` value while attempting to fetch the route Url { generic_url: "http://localhost:3000/regex/latest/regex/" }

Stack backtrace:
   0: anyhow::private::new_adhoc
             at /home/jsha/.cargo/registry/src/github.com-1ecc6299db9ec823/anyhow-1.0.42/src/lib.rs:632:36
   1: docs_rs::web::rustdoc::rustdoc_html_server_handler
             at ./src/web/rustdoc.rs:331:17
   2: core::ops::function::Fn::call
             at /rustc/1af55d19c7a9189374d89472f97dc119659bb67e/library/core/src/ops/function.rs:70:5
   3: <F as iron::middleware::Handler>::handle
             at /home/jsha/.cargo/registry/src/github.com-1ecc6299db9ec823/iron-0.6.1/src/middleware/mod.rs:389:9
   4: <alloc::boxed::Box<dyn iron::middleware::Handler> as iron::middleware::Handler>::handle
             at /home/jsha/.cargo/registry/src/github.com-1ecc6299db9ec823/iron-0.6.1/src/middleware/mod.rs:395:9
   5: <docs_rs::web::metrics::RequestRecorder as iron::middleware::Handler>::handle
             at ./src/web/metrics.rs:50:22
   6: <alloc::boxed::Box<dyn iron::middleware::Handler> as iron::middleware::Handler>::handle
             at /home/jsha/.cargo/registry/src/github.com-1ecc6299db9ec823/iron-0.6.1/src/middleware/mod.rs:395:9
   7: <alloc::boxed::Box<dyn iron::middleware::Handler> as iron::middleware::Handler>::handle
             at /home/jsha/.cargo/registry/src/github.com-1ecc6299db9ec823/iron-0.6.1/src/middleware/mod.rs:395:9
   8: router::router::Router::handle_method
             at /home/jsha/.cargo/registry/src/github.com-1ecc6299db9ec823/router-0.6.0/src/router.rs:199:18
   9: <router::router::Router as iron::middleware::Handler>::handle
             at /home/jsha/.cargo/registry/src/github.com-1ecc6299db9ec823/router-0.6.0/src/router.rs:212:9
  10: <alloc::boxed::Box<dyn iron::middleware::Handler> as iron::middleware::Handler>::handle
             at /home/jsha/.cargo/registry/src/github.com-1ecc6299db9ec823/iron-0.6.1/src/middleware/mod.rs:395:9
  11: iron::middleware::Chain::continue_from_handler
             at /home/jsha/.cargo/registry/src/github.com-1ecc6299db9ec823/iron-0.6.1/src/middleware/mod.rs:359:15
  12: iron::middleware::Chain::continue_from_before
             at /home/jsha/.cargo/registry/src/github.com-1ecc6299db9ec823/iron-0.6.1/src/middleware/mod.rs:353:9
  13: <iron::middleware::Chain as iron::middleware::Handler>::handle
             at /home/jsha/.cargo/registry/src/github.com-1ecc6299db9ec823/iron-0.6.1/src/middleware/mod.rs:277:9
  14: <alloc::boxed::Box<dyn iron::middleware::Handler> as iron::middleware::Handler>::handle
             at /home/jsha/.cargo/registry/src/github.com-1ecc6299db9ec823/iron-0.6.1/src/middleware/mod.rs:395:9
  15: <docs_rs::web::MainHandler as iron::middleware::Handler>::handle::{{closure}}::{{closure}}
             at ./src/web/mod.rs:179:39
  16: <docs_rs::web::MainHandler as iron::middleware::Handler>::handle::if_404
             at ./src/web/mod.rs:158:17
  17: <docs_rs::web::MainHandler as iron::middleware::Handler>::handle::{{closure}}
             at ./src/web/mod.rs:179:26
  18: core::result::Result<T,E>::or_else
             at /rustc/1af55d19c7a9189374d89472f97dc119659bb67e/library/core/src/result.rs:1032:23
  19: <docs_rs::web::MainHandler as iron::middleware::Handler>::handle
             at ./src/web/mod.rs:177:9
  20: <iron::iron::RawHandler<H> as hyper::server::Handler>::handle
             at /home/jsha/.cargo/registry/src/github.com-1ecc6299db9ec823/iron-0.6.1/src/iron.rs:176:17
  21: hyper::server::Worker<H>::keep_alive_loop
             at /home/jsha/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.10.16/src/server/mod.rs:340:13
  22: hyper::server::Worker<H>::handle_connection
             at /home/jsha/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.10.16/src/server/mod.rs:282:15
  23: hyper::server::handle::{{closure}}
             at /home/jsha/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.10.16/src/server/mod.rs:242:34
  24: hyper::server::listener::spawn_with::{{closure}}
             at /home/jsha/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.10.16/src/server/listener.rs:50:31
  25: std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/1af55d19c7a9189374d89472f97dc119659bb67e/library/std/src/sys_common/backtrace.rs:123:18
  26: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
             at /rustc/1af55d19c7a9189374d89472f97dc119659bb67e/library/std/src/thread/mod.rs:483:17
  27: <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/1af55d19c7a9189374d89472f97dc119659bb67e/library/core/src/panic/unwind_safe.rs:271:9
  28: std::panicking::try::do_call
             at /rustc/1af55d19c7a9189374d89472f97dc119659bb67e/library/std/src/panicking.rs:403:40
  29: __rust_try
  30: std::panicking::try
             at /rustc/1af55d19c7a9189374d89472f97dc119659bb67e/library/std/src/panicking.rs:367:19
  31: std::panic::catch_unwind
             at /rustc/1af55d19c7a9189374d89472f97dc119659bb67e/library/std/src/panic.rs:133:14
  32: std::thread::Builder::spawn_unchecked::{{closure}}
             at /rustc/1af55d19c7a9189374d89472f97dc119659bb67e/library/std/src/thread/mod.rs:482:30
  33: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/1af55d19c7a9189374d89472f97dc119659bb67e/library/core/src/ops/function.rs:227:5
  34: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/1af55d19c7a9189374d89472f97dc119659bb67e/library/alloc/src/boxed.rs:1691:9
  35: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/1af55d19c7a9189374d89472f97dc119659bb67e/library/alloc/src/boxed.rs:1691:9
  36: std::sys::unix::thread::Thread::new::thread_start
             at /rustc/1af55d19c7a9189374d89472f97dc119659bb67e/library/std/src/sys/unix/thread.rs:106:17
  37: start_thread
             at ./nptl/./nptl/pthread_create.c:435:8
  38: __clone
             at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone.S:100
```

jyn514 replied:
Oh err hmm, probably lots of routes are feeding it into a database query. Not sure what the proper fix is, I'd have to think about it. I'd be ok with fixing the routes one at a time if you think it's easier than changing match_version.

jsha replied:
Great, thanks! I've uploaded a version that fixes all the routes. Here's a quick-and-dirty script that demonstrates the URLs we care about:

```
(
curl -s http://localhost:3000/regex/latest/regex -i
curl -s http://localhost:3000/regex/ -i
curl -s http://localhost:3000/regex -i
curl -s http://localhost:3000/crate/regex -i
curl -s http://localhost:3000/crate/regex/latest -i
curl -s http://localhost:3000/crate/regex/latest/ -i
) | grep Location
```

```
Location: http://localhost:3000/regex/latest/regex/
Location: http://localhost:3000/regex/latest/regex/
Location: http://localhost:3000/regex/latest/regex/
Location: http://localhost:3000/crate/regex/latest
Location: http://localhost:3000/crate/regex/latest
```

If the code looks like it's heading in the right direction I can go ahead and write a unittest.

jyn514 replied:
Yup, that looks good :) and I thought of the proper "fix" - add a new variant that's `MatchSemver::Latest` and includes the version from the database, that way you can't forget to handle it. (e.g. I suspect you forgot to handle docs.rs' source view)

jsha commented:
Alright, I've added test cases for the crate page redirects, and the doc redirects are covered by the existing test cases, which I updated. Ready for re-review.

jyn514 commented:
@jsha do you mind implementing my `MatchVersion::Latest` suggestion https://github.com/rust-lang/docs.rs/pull/1527#discussion_r739613990 ? Sorry to keep changing my mind, this should be the last time hopefully.

jsha commented:
Alright, re-did with MatchSemver::Latest. That was less work than I had worried it would be, and you're right that it's tidier.

jyn514 commented on src/web/source.rs:
```diff
@@ -179,6 +179,7 @@ pub fn source_browser_handler(req: &mut Request) -> IronResult<Response> {
         crate_name = new_name;
     }
     let version = match v.version {
+        MatchSemver::Latest((version, _)) => version,
         MatchSemver::Exact((version, _)) => version,
```
```suggestion
        MatchSemver::Exact((version, _)) | MatchSemver::Latest((version, _)) => version,
```

jsha commented on src/web/features.rs:
```diff
@@ -259,7 +260,7 @@ mod tests {
 
             let resp = env.frontend().get("/crate/foo/latest/features").send()?;
-            assert!(resp.url().as_str().ends_with("/crate/foo/0.2.0/features"));
+            assert!(resp.status().is_success());
```
That's a bit odd since the goal is to test that this serves immediately (200, ~= is_success) rather than redirects (302). The version you propose would pass if we redirected to self (or to any other path that ends with the requested URL)

jyn514 replied:
@jsha `reqwest.get()` automatically follows redirects. If you want to test that there's no intermediate redirects, you should call `override_frontend`: https://github.com/rust-lang/docs.rs/blob/495e56c92a827888f4ae78a05e35f7e639f39079/src/web/rustdoc.rs#L1032-L1037

Once you set it not to follow redirects, I still think it makes sure to test the path directly to make sure it's going to the right page.

jsha commented on src/web/mod.rs:
```diff
@@ -351,7 +355,11 @@ fn match_version(
         return Ok(MatchVersion {
             corrected_name,
-            version: MatchSemver::Semver((version.to_string(), *id)),
+            version: if input_version == Some("latest") {
```
I thought what we discussed above was that `*` and `newest` should continue to have their current behavior. That's why I specifically went for `input_version` rather than `req_version` (which gets set to `*` for all three cases).

jyn514 replied:
My impression was they should redirect to *latest*, not to the exact version. It seems weird to have that be inconsistent.

(But uhhh yeah I did forget they should redirect, oops)

jsha replied:
In what context are `*` and `newest` exposed? Part of why I wanted to leave these as-is is that I don't fully understand the use case for these alternate names. Are they a fully supported feature? How do people get to them?

I _think_ supporting these by redirecting to `latest` will require an additional clause in each of the places we redirect, so I want to make sure this is something we want to do before going ahead with it.

jyn514 replied:
Yes, they are fully supported in that we document them on https://docs.rs/about/redirections.

jsha replied:
I looked at fixing this and it complicates the code somewhat. We'd have to add a _fourth_ MatchSemver variant, RedirectToLatest or similar. I propose to let `*` and `newest` continue to redirect to an exact version. It seems slightly weird, but it has the advantage of continuing the current behavior and not overcomplicating the code (and being less work :-D).

jyn514 replied:
Hmm, ok - it doesn't seem like the worst thing to have a way to opt-in to the old behavior :)

jsha commented on src/web/builds.rs:
```diff
@@ -329,19 +330,13 @@ mod tests {
-                .url()
-                .as_str()
-                .ends_with("/crate/aquarelle/0.2.0/builds.json"));
+            assert!(resp_json.status().is_success());
```
Like above, I think testing for a 200 response (i.e. not a redirect) is the right thing here.

jyn514 commented:
Hi @jsha, do you know when you'll get a chance to look at this again? It's almost done I think, it just needs a few updates to the tests :) (https://github.com/rust-lang/docs.rs/pull/1527#discussion_r748725526)

jsha commented:
Thanks for the ping! I'll try to work on this this weekend. In addition to the tests there's also a pending request for a change in functionality (`newest` and `*` to redirect to `latest` instead of a specific version): https://github.com/rust-lang/docs.rs/pull/1527#discussion_r739744467

jsha commented:
Updated and ready for re-review, pending resolution of https://github.com/rust-lang/docs.rs/pull/1527#discussion_r739744467.

jyn514 commented:
@jsha Can you also update the crate details page to point to `/latest` unconditionally? Right now they point to the latest version number instead, these tabs:
![image](https://user-images.githubusercontent.com/23638587/143438704-4bed3452-2f0a-40ab-ba6e-dffb3a2ba4c8.png)

the "platform" and "details" dropdowns are also using version numbers:
![image](https://user-images.githubusercontent.com/23638587/143438886-daf0ccb6-12fe-4dda-8a3f-348dd1abca57.png)

also if you don't mind it would be nice to remove the merge commit by rebasing, but I don't feel strongly about that.

jsha commented:
Will do. Also there's a similar issue going from the doc page, via the menu, to the crate page - if you're on `/latest/`, it will send you to a specific version.

To clarify: you said "crate details page to point to /latest unconditionally," but I think the right logic is "all links to versioned URLs point to /latest if the current page is /latest, or the version of the current page otherwise." Do you agree?

Regarding rebases: My general preference is to squash everything into one commit before merge, and also to not rebase or squash mid-review, so reviewers can easily see the diffs between revisions. In other words, please never hesitate to ask me to rebase, I'm happy to do it. I'm also of course happy to adapt to a "rebase after each round of feedback" approach if that works better for you.

jyn514 commented:
> To clarify: you said "crate details page to point to /latest unconditionally," but I think the right logic is "all links to versioned URLs point to /latest if the current page is /latest, or the version of the current page otherwise." Do you agree?

Oh oops, good catch, that's what I meant :)

> My general preference is to squash everything into one commit before merge, and also to not rebase or squash mid-review, so reviewers can easily see the diffs between revisions.

Sounds great, thanks!

jsha commented:
Pushed some fixes to the crate page and friends. That wound up being more involved than I expected, so please take a close look. I added some tests for those behaviors.

I also added a "Permalink" under the crate menu to link to a versioned URL.

jyn514 commented:
> I also added a "Permalink" under the crate menu to link to a versioned URL.

Ooh, that reminds me - github/gitlab have a `y` keyboard shortcut which changes the current URL to have the version number, could we do that here too? In a follow-up PR is fine, I know this one is getting kind of big.

The last thing I think we should change is for "go to latest version" to use `/latest/` as well.

The /target-redirect/ page doesn't seem to support /latest/:
![image](https://user-images.githubusercontent.com/23638587/143683081-0bff21a6-78cf-478c-998e-b68fd89e6451.png)
/crate/{crate_name}/latest/builds/184 also is broken.

(I haven't looked at the code yet, I'll do that in a second.)

jyn514 commented on src/web/rustdoc.rs:
```diff
@@ -447,11 +447,8 @@ pub fn rustdoc_html_server_handler(req: &mut Request) -> IronResult<Response> {
+    let mut latest_path = if latest_release.build_status {
         let target = if target.is_empty() {
             &krate.metadata.default_target
         } else {
```
This is around the code to modify for "go to latest", I think you can just switch `latest_version` to hard-coding `"latest"` below.

jsha replied:
I actually need to keep this code, or something like it, because I use it to calculate the target of the `Permalink` URL when the user is on a `/latest/` URL.

jyn514 reviewed (approved):
Code looks great :) thanks for writing so many tests!

jsha commented:
> Ooh, that reminds me - github/gitlab have a y keyboard shortcut which changes the current URL to have the version number, could we do that here too? In a follow-up PR is fine, I know this one is getting kind of big.

Yep, let's do this in a followup PR.

> The last thing I think we should change is for "go to latest version" to use /latest/ as well.

I'd like to propose also doing this in a followup PR. It's fairly isolatable and is in the general category of "start directing more traffic to the /latest/ URLs," along with updating the sitemap and changing crate.io.

I pushed some commits, but I just realize my fix for the builds page is incorrect (goes to a versioned page, but that's not actually what we want). I'll work on that some more and push again later today.

jyn514 commented:
> I pushed some commits, but I just realize my fix for the builds page is incorrect (goes to a versioned page, but that's not actually what we want). I'll work on that some more and push again later today.

Hmm, no that seems correct to me - we want to show a permalink because people often copy/paste it into a github issue.

jyn514 commented:
This looks good to me as soon as you squash and fix the clippy lint :) thank you so much for sticking with this!
